src/Pate/Proof.hs

```diff
@@ -143,82 +136,100 @@ instance PEM.ExprMappable sym (EquivTripleBody sym arch) where
+  mapExpr sym f (InequivalenceResult a1 a2 a3 a4) = InequivalenceResult
+    <$> PEM.mapExpr sym f a1
+    <*> PEM.mapExpr sym f a2
+    <*> PEM.mapExpr sym f a3
+    <*> pure a4
+
+type InequivalenceResult arch = PG.Grounded (InequivalenceResultC arch)
+
+withIneqResult ::
+  InequivalenceResult arch ->
+  (forall grnd. PA.ValidArch arch => PG.IsGroundSym grnd => InequivalenceResultC arch grnd -> a) ->
+  a
+withIneqResult ineq f = PG.withGroundSym ineq $ \ineq'@InequivalenceResult{} -> f ineq'
+
+data CondEquivalenceResult sym arch where
+  CondEquivalenceResult :: PA.ValidArch arch =>
+    { condEqExample :: MapF.MapF (W4.SymExpr sym) W4G.GroundValueWrapper
+    -- ^ bindings for variables in counter-example
+    , condEqPred :: W4.Pred sym
+    -- ^ condition that is sufficient to imply equivalence
+    , condEqAbortValid :: Bool
+    -- ^ true if the negation of this predicate necessarily implies an
+    -- abort path on the original binary
+    } -> CondEquivalenceResult sym arch
+
+emptyCondEqResult ::
+  forall arch sym.
+  PA.ValidArch arch =>
+  PSo.ValidSym sym =>
+  sym ->
+  CondEquivalenceResult sym arch
+emptyCondEqResult sym = CondEquivalenceResult MapF.empty (W4.falsePred sym) False
+
+
+
+instance PEM.ExprMappable sym (CondEquivalenceResult sym arch) where
+  mapExpr _sym f (CondEquivalenceResult e1 e2 b) =
+    CondEquivalenceResult
+      <$> (MapF.fromList <$> traverse (\(MapF.Pair e1' e2') -> MapF.Pair <$> f e1' <*> pure e2') (MapF.toList e1))
+      <*> f e2
+      <*> pure b
+
+
+-- | The status of some verification problem: either producing a successful result
+-- or yielding a counterexample. The resulting
+-- predicate represents an additional assumption required for equivalence to hold.
+-- (i.e. conditional equivalence).
+-- In the case of inequality, the predicate simply false
+data VerificationStatus sym arch =
     Unverified
   | VerificationSkipped
   | VerificationSuccess
-  | VerificationFail ce
-  deriving (Functor, Foldable, Traversable)
-
-instance PEM.ExprMappable sym a => PEM.ExprMappable sym (VerificationStatus a) where
-  mapExpr sym f = traverse (PEM.mapExpr sym f)
-
-data ProofFunctionCallMetadata where
-  ProofFunctionCallMetadata ::
-    { prfFunctionCallMetadataAddress :: PA.ConcreteAddress arch
-    } -> ProofFunctionCallMetadata
+  | VerificationFail
+      { verStatusCounterExample :: InequivalenceResult arch,
+        verStatusCondEq :: CondEquivalenceResult sym arch
+      }
+
+instance PEM.ExprMappable sym (VerificationStatus sym arch) where
+  mapExpr sym f s = case s of
+    VerificationFail ce condeq -> do
+      ce' <- PEM.mapExpr sym f ce
+      condeq' <- PEM.mapExpr sym f condeq
+      return $ VerificationFail ce' condeq'
+    _ -> return s
 
 -- | An abstract proof object, representing the overall structure of an equivalence proof.
 -- The type parameter 'prf' abstracts this structure over the specific types used in the proof,
```

Comment:
I think this comment might need to be updated now that the type parameters have changed